### PR TITLE
add traits to use echo broadcasting

### PR DESCRIPTION
**Pull request by patrickweh**

### CODE REVIEW

File: src/Traits/BroadcastsEvents.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+        return $instance->broadcastChannel($generic);
+    }
+
+    protected function newBroadcastableEvent($event): BroadcastableModelEventOccurred
+    {
+        return (new BroadcastableModelEventOccurred($this, $event))->dontBroadcastToCurrentUser();
+    }
+
+    public function broadcastOn($event): array
```
Comment (SirSplasch) [inline src/Traits/BroadcastsEvents.php:49]: public function broadcastOn($event): array
    {
        return [new PrivateChannel($this->broadcastChannel($event === 'created'))];
    }

Comment (SirSplasch) [reply src/Traits/BroadcastsEvents.php:49]: Keep it as simple and generic as possible. If someone needs related broadcasting, one can extend this function to fulfill their own needs.

File: src/Traits/HasEloquentListeners.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    public function getPaginator(LengthAwarePaginator $paginator): LengthAwarePaginator
+    {
+        $this->broadcastChannels = [];
+        foreach ($paginator->items() as $item) {
+            $this->broadcastChannels[$item->getKey()] = $item->broadcastChannel();
+        }
+
+        if (
+            in_array(BroadcastsEvents::class, class_uses_recursive($this->model))
+            && $paginator->currentPage() === 1) {
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:20]: break ") {" into a new line

File: src/Traits/HasEloquentListeners.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            array_pop($this->data['data']);
+        }
+
+        $this->skipRender();
+    }
+
+    public function echoDeleted($eventData): void
+    {
+        $data = \Arr::keyBy($this->data['data'], $this->modelKeyName);
+        unset($data[$eventData['model'][$this->modelKeyName]]);
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:66]: no need for 2 unset statements, can be done in one.

File: src/Traits/HasEloquentListeners.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        $this->data['data'] = array_values($data);
+        unset($this->broadcastChannels[$eventData['model'][$this->modelKeyName]]);
+        $this->data['total'] = $this->data['total'] - 1;
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:69]:
```suggestion
        --$this->data['total'];
```

File: src/Traits/HasEloquentListeners.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        $this->data['to'] = $this->data['to'] - 1;
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:70]:
```suggestion
        --$this->data['to'];
```

File: src/Traits/HasEloquentListeners.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        $this->skipRender();
+    }
+
+    public function echoCreated($data): void
+    {
+        $model = $this->getBuilder($this->model::query()->whereKey($data['model'][$this->modelKeyName]))->first();
+        $item = $this->itemToArray($model);
+
+        array_unshift($this->data['data'], $item);
+        $this->data['total'] = $this->data['total'] + 1;
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:52]:
```suggestion
        ++$this->data['total'];
```

File: src/Traits/HasEloquentListeners.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        $this->data['to'] = $this->data['to'] + 1;
```
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:53]:
```suggestion
        ++$this->data['to'];
```

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: src/Traits/HasEloquentListeners.php (hunk 6 above)
Comment (SirSplasch) [inline src/Traits/HasEloquentListeners.php:52]:
```suggestion
        $this->data['total']++;
```

File: src/Traits/BroadcastsEvents.php (hunk 1 above)
Comment (patrickweh) [reply src/Traits/BroadcastsEvents.php:49]: √

File: src/Traits/HasEloquentListeners.php (hunk 2 above)
Comment (patrickweh) [reply src/Traits/HasEloquentListeners.php:20]: √

File: src/Traits/HasEloquentListeners.php (hunk 3 above)
Comment (patrickweh) [reply src/Traits/HasEloquentListeners.php:66]: √

**Review by SirSplasch — APPROVED** [review]

**Pull request merged**